python/phase-2.md

```diff
@@ -0,0 +1,996 @@
+
+## Runtime
+
+### Workflow Stepped Execution
+
+Notes:
+
+* We are gonna use `asyncio`, but whether we have to use our own event loop remains to be seen
+  * Have shown at https://github.com/cretz/python-determinism-runner basically using the approach from
+    https://stackoverflow.com/a/29797709 that we can run a single step and enforce that we have yielded only at
+    workflow-acceptable spots
+* Do want to prevent use of `asyncio.sleep`, `asyncio.wait_for`, and similar that are based on system time?
+  * This is much clearer than replacing their implementations so that people aren't confused about why a 50ms wait does
+    not work
+  * We may not be able to block everything until we use a sandbox (see next section)
+  * Making alternatives is a tad ugly, e.g. one can use `asyncio.gather()` without timeout but they can't with timeout,
+    instead they'd have to have a sleep or something
+
+### Workflow Deterministic Execution
+
+Notes:
+
+* A full workflow execution sandbox may not occur until phase 3
+  * Some work on doing this via `eval` are at https://github.com/cretz/python-determinism-runner
+* We will, however, go ahead and abstract the "runner" so that we can make it pluggable/opt-out if the caller chooses
+
+## Additional API Concerns
+
+### Activities as Methods
+
+It is very normal to want to have activities as methods instead of top-level functions. For example, say you have:
+
+```python
+class ActivityClass:
+    def __init__(self, some_init_arg: int) -> None:
+        pass
+    
+    @activity.defn
+    def some_activity(self, arg1: str) -> List[str]:
+        return ["Yay"]
+```
+
+You cannot just `workflow.execute_activity(ActivityClass.some_activity, "foo")` because technically the first parameter
+is `self` there. But sometimes they may want to call
+`workflow.execute_activity(ActivityClass.some_activity, ActivityClass("init-arg1"), "foo")` and who are we to say they
+can't pass in the self arg explicitly?
+
+What do we do? Here are the notes/options (activity options left off of calls for brevity):
+
+* Cannot use something like `functools.partial(ActivityClass.some_activity, None)` because `partial` removes our
+  attributes which are set in the decorator
+* Don't want to init an instance of the class like this:
+
+    ```python
+    workflow.execute_activity(ActivityClass("init-arg").some_activity, "foo")
+    ```
+* Can create the activity class without initializing it like this:
+
+    ```python
+    activities = ActivityClass.__new__(ActivityClass)
+    workflow.execute_activity(activities.some_activity, "foo")
+    ```
+  This is nice because it doesn't require the Python SDK to care (and people have been doing this in class methods for
+  factories for a while now). We can even provide a helper that does basically this like so:
+
+    ```python
+    activities = workflow.proxy_activity_class(ActivityClass)
+    workflow.execute_activity(activities.some_activity, "foo")
+    ```
```

Comment:
This is probably my favorite option in this list but I still don't completely love it.

Pretty sure you already said this but there's really no way to have the decorator detect and strip the self param for usage specifically in this situation when being called from a workflow? It's a bit like what you suggest with wrapping the function below, except without needing to have `defn` and `method_defn` be different things. If we did that, I'm worried people would forget constantly and use the wrong one (although, ideally we could just blow up at init time if they did)

Reply:
> This is probably my favorite option in this list but I still don't completely love it.

I too don't like it, but there is no perfect option at the moment to separate top-level functions from methods

> there's really no way to have the decorator detect and strip the self param for usage specifically in this situation when being called from a workflow?

Not at compile time because the decorator at compile time can't differentiate between method w/ "self" and top-level function with just a param. They just look like params at compile time (at runtime I can inspect and determine it's a method instead of a top-level function).

> If we did that, I'm worried people would forget constantly and use the wrong one

Concur